### ISSUE TITLE
[Discuss] The future of the classic notebook interface

### ISSUE THREAD

**Issue by Zsailer** [jupyter/notebook#6210]
The purpose of this issue is to discuss the future of the jupyter/notebook repository. 

**Background:** The classic Jupyter notebook interface is currently maintained by a skeleton crew of developers. On the other hand, many people still use the classic interface today. 

**Problem:** We have a widely-used piece of software with very little developer support. 

Let's have an open discussion about what to do with this project moving forward.

**Comment by Zsailer** [comment]
An alternative proposal was to offer this project as an opportunity for another team/company/group to maintain this project moving forward. There might be some team with vested interest in making sure this project stays alive indefinitely. 

There was some concern raised with this idea. Handing off the project to another (possibly not Jupyter-affiliated) team puts the Jupyter brand (?, just echoing what I've heard said here) at risk.

**Comment by meeseeksmachine** [comment]
This issue has been mentioned on **Jupyter Community Forum**. There might be relevant details there:

https://discourse.jupyter.org/t/the-future-of-classic-notebook-interface/11328/1

**Comment by psychemedia** [comment]
*Apologies in advance for a long post... This repeats a lot of things posted elsewhere, but I think I still haven't found a way to express the growing gulf I see between the experience and expectations of highly skilled developers, and people who are very naive end-users of Jupyter tools doing very, very simple tasks that appear to be little more than (and often are very little more than) the following to a seasoned developer:*

```
for i in [1, 2, 3, 4]:
  print("hello world")
```

*At the same time, this is not to say that we shouldn't be providing such novices with environments that are, underneath, very powerful, and scalable in terms of what you can do with them. What I try to do below is explore what notebooks made possible, who they empowered, and what we are at risk of losing.*

As I've commented on this issue many times, I'll argue the position of:

- an educator, using notebooks to present material to novice learners in disciplines that may only touch tangentially on computing (for example, a newly graduated K12 teacher teaching basic programming to a class of two dozen 12 year olds; a near-retirement lecturer teaching first year undergrad mechanical engineering);
- educator coding experience: basic HTML, inline style/CSS, maybe some elementary (non-framework) Javascript (2 -30 lines enough to implement a basic function to do a single thing);
- institutional support: essentially none; maybe some grudging hosting of a JupyterHub server.

Skill level of users: JupyterLab IDE is overkill. Classic notebook UI is at the limit of UI complexity; notebook extension toolbar buttons bewilder a significant a percentage of users and make the UI unusable.

Duty cycle for upgrading software: 5 years +. (In my institution, modules run for 5-6 years then possibly get extended for another 3+ years. They take two years to produce. The software freeze (except for minor updates once a year) is typically during the first year of production, so things then need to stay largely in place for the next 7-8 years. In general, educational materials get reused year on year on year. Computers in K12 schools may be 5 years+ old. Note that education needs are global and extend to developing nations not just Ivy League feeder High Schools and inner-city deprived ghetto schools with low student expectations and harassed teaching staff.

The original notebooks allowed people who could write basic HTML pages to write extensions; the [`jupyter_contrib_nbextensions`](https://github.com/ipython-contrib/jupyter_contrib_nbextensions) package includes many extensions created very early on. Many of those packages were based around very simple basic javascript and served as a template for creating new extensions based on copying code. Looking at the same extensions now (with frameworks and promises and syntax and far-from basic Javascript), I would struggle to create my own extension based on inspection of that code. And JupyterLab extension code is an order of magnitude harder to get starter with in terms of code complexity, required knowledge of Jupyter framework/API calls, and development environment.

My skill level excludes me from updating my original classic notebook extensions. There is no internal developer support available eg for porting extensions to JupyterLab. I do not have the skill or the time to learn either js frameworks or js dev tools, let alone make sense of JupyterLab API docs and how-to example fragments that are buried in code tests. My instructional material relies on classic Jupyter notebooks with their current extensions.

My institution has large numbers of developers working in institutional IT projects and growing numbers of courses using Jupyter notebooks, but with zero central IT or developer support provided to support the development of educational end-use of Jupyter notebooks. At the moment, pretty much all "development" is carried out by academics who are not developers (why should a statistician using LaTex and R know how to develop front-end web apps?) and who are supposed to be spending their time developing subject matter material and teaching students. For them, the Jupyter notebook environment is a means to an end - a way of maximising the chance that a novice with very poor computing skills can run 3 lines of code and see the result. The notebook model means the code can be placed in instructional text and made into an interactive textbook.

In many cases, Jupyter Book + Thebe (and even better, thebe w/ a pyodide kernel) would be appropriate for supporting this sort of interactive text book use case. But it lacks the ability to persist code changes or add and persist new cells against a user account.

In other cases, the more general authoring environment that classic notebook allows users to populate their own notebook (a digital version of a kid's paper school exercise book that they write their homework in), with text, media assets, maybe some code and code outputs (with a total single lines of code count across the whole notebook in the low tens *at most*). A single notebook might take up a complete K12 school lesson (35 minutes) and be the equivalent of a page or two pages of A4 paper.

My fear is that this large body of legacy users, who you are unlikely to see on Github, are going to be left in limbo (why would a K12 maths teacher in a rural school who has a dozen or so notebooks used over a whole school term to run some simple ipyturtle activities, need a Github account?). I think we saw some signs of this when a recent classic notebook update clobbered inline HTML styles in markdown cells.

Classic notebooks provided folk who don't have access to webservers, and don't need the grief of installing or supporting even basic, default Python environments, with a read-write-render-execute medium for HTML, rich media (embedded images, videos, audio), and REPL based code execution and code output previewing. Teachers of 12 year olds could get instructional material they'd written in front of kids, and the kids could run the code, write their own, write their own notes, and save a copy of it.

That's what drove my adoption. Classic notebook was an on-ramp. If I came to Jupyter now, and was exposed to JupyterLab as the first experience, I'd probably walk away because I'd stand zero chance of persuading anyone else it was something we'd be able to use with very novice distance education students. (It's been hard enough selling classic notebooks.)

Another fear is that a company like Pearson will pick up the maintenance of classic notebook and then lock users in. A regret is that institutions such as my own (which contributes to Moodle core development, for example) do not appear to value the benefit of putting support into classic notebook, not in terms of improving accessibility, nor providing maintenance effort, nor helping turn educator contributed have-a-go extensions into production stable extensions that can be of general effort.

The Jupyter project is a phenomenal undertaking, but it's evolution is starting to mirror that of the web. A couple of decades a go, you could see something interesting on a web page, view source, and stand a good chance of being able to appropriate the HTML+CSS code you saw and do something interesting. Classic notebook provided a UI that supported that level of engagement. But it's not what developers who want to push things forward are interested in. We need to find a way of resourcing some developers who want to push things back to as simple an environment as possible, that still affords the power of working with other Jupyter components.

IMVHO!;-)

**Comment by Carreau** [comment]
Today's governance call was used to discuss many of the above topics. It just finished a few minutes ago, and we're going to posts the notes we took during this call after a bit of necessary cleanup.

**Comment by cezarymierzejek** [comment]
> > Things like being able to link notebooks and consoles to the same kernel, viewing the same file in multiple ways, git integration, debugging capabilities, ability to view multiple files (and file types) at the same time provide enormous leaps in productivity.
> 
> I think that does relate to the different use-cases that you've indentified. I don't know, but you may have identified the use-case JupyterLab was particularly desiged for - a data scientist who spends does a lot of their work in notebooks
> 
> And I think you're right - that's not what I do as a teacher - or as an experienced Python coder. From talking to other people - including Jupyter developers - I think I'm reasonably typical - in that I use notebooks as sketches, and as tutorials and demos. I often buff-up a notebook sketch by converting to RMarkdown, going into my text editor and spending a lot of time editing there. I very rarely debug in the notebook, because if the code is at all substantial, I switch to my text editor again, and put the code into a `.py` files where I can edit them more easily, write tests, use standard version control. I find the notebook very uncomfortable for working with any significant amount of code, and I have trained myself to switch out of the notebook quickly when I realize that is happening.
> 
> For those cases - the notebook we need is something much lighter. It's not meant to be a platform for your coding life, but a gateway to other things. But a gateway that is of enormous use and benefit. I think you're right to draw that distinction, and you can see it everywhere in this thread, but particularly from comments by @psychemedia and @Carreau.

Completely agree with your comments re: the different use cases, users, and what they are looking for. 

I think part of what is fueling the debate on what direction JupyterLab (or Jupyter in general) needs to take is that it typically operates in close proximity with other tools in many data science/data analytics workflows. This is in contrast to what you see in similar communities (e.g., RStudio, Stata, Matlab) where typically there is 1 tool/tool chain to rule them all.

Most people I see pair JupyterLab with either VS Code or PyCharm for the exact reasons that you described - as certain pieces of code or logic get matured in a notebook it eventually makes more sense to transition and package them into .py files. 

What I think has changed is that these adjacent tools, which largely ignored notebooks in the past and were primarily used in this transition and packaging step, are now trying to take a bigger bite of the notebook workflow. VS Code and PyCharm being the most prominent examples. 

I personally favor the “platform for your coding life” that you mentioned but can understand that is a very unwelcome direction for those who are looking for a more lightweight tool. Optimistically, I think projects like retrolab look like they can get there and hopefully a solution can be found for those critical extensions only available for classic notebooks.

**Comment by matthew-brett** [comment]
> I personally favor the “platform for your coding life” that you mentioned but can understand that is a very unwelcome direction for those who are looking for a more lightweight tool.

Well - also - I really wonder whether it is realistic for JupyterLab to compete in that space, when VSCode - for example - is already so prominent there, and has shown so much interest in developing tools for working with notebooks.

**Comment by cezarymierzejek** [comment]
> > I personally favor the “platform for your coding life” that you mentioned but can understand that is a very unwelcome direction for those who are looking for a more lightweight tool.
> 
> Well - also - I really wonder whether it is realistic for JupyterLab to compete in that space, when VSCode - for example - is already so prominent there, and has shown so much interest in developing tools for working with notebooks.

It’s a fair question. 

I feel more optimistic that it can because in my experience VS Code’s implementation of notebook capabilities (at least the last time I tried it a couple of months ago) are quite buggy.    

On paper, VS Code appears to have replicated many features of classic notebook/JupyterLab or have been faster to implement a subset of long standing asks (e.g., variable viewer like RStudio). In practice, they break / crash quite often when working with datasets of any decent size. e.g., 1 GB or bigger (which I consider quite small). 

It’s possible that VS Code will fix many of these issues eventually but I think that is a big if given that its primary audience is developers working with text files. I wonder how much fixing these issues requires core changes in VS Code and potentially degrading their experience vs what is possible to do as an extension (e.g., popular features like shift tab completion being open issues for 2+ years makes me believe it is harder to copy what Jupyter can do than it looks) 

My guess is that these limitations are part of the reason why they are focused on the volume of features e.g., try to match 100% of what classic/lab can do even if they are only 50% as good in practice. The benefit of this approach is that they can scoop up the jupyter users who are running very simple notebook workflows, but also represent a big population. 

Overall, I think that there is sufficient difference in use-cases between doing data/analytical work with pure development work that I think JupyterLab can compete. And at bigger data scales these differences are much more substantial.

**Comment by rickmcgeer** [comment]
As I read the many thoughtful posts and comments on this thread, it seems clear that the primary attraction of the classic Notebook interface is the large body of classic Notebook extensions that don't work, as yet, in JupyterLab tabs.  A second attraction is that it was very easy for Notebook users to extend the classic Notebook without rebuilding the Notebook editor, which currently can't be done in JupyterLab.

For any complex system such as Jupyter, there are three levels of extension:

- _Customization_, typically done through modifying static settings.
- _User Extensions_, which are done without modifying the underlying system (a quick test here is whether the underlying system needs to be recompiled in order to incorporate the extension).  The early productivity tools did this through macro languages and Visual Basic.  Scripting languages (including Python) arose as a way to rapidly customize software systems through user extensions.  In fact, Guido van Rossum in Python followed John Ousterhout in Tcl in thoroughly and carefully designing the "southbound" interface for his language.
- _Developer Extensions_, which require recompilation and rebuild of the full system in order to implement the extension.

It's important to note that these aren't exclusive.  Software systems can and do incorporate all three in a single system.  Simple extensions are done at the scripting layer as user extensions; more complex ones are implemented as Developer Extensions.  In fact, it's not uncommon for a single extension to begin life as a User Extension, and then (for ease of maintenance, performance, or robustness) be re-implemented as a Developer Extension when mature.

JupyterLab has a very well-designed and carefully-implemented Developer Extension system, and the many extensions that have been developed for JupyterLab are testament to its power and value.   We have built one that opens a visual editor in a JupyterLab tab and interacts with the filesystem through JupyterLab as a user manipulates objects in the editor.  We aren't the only people to do that, of course -- the Vega extension does similar things.

However, it appears that the User Extension system is less well-developed than the Developer Extension system in JupyterLab.  For example, if we look at the extension tutorial in the JupyterLab documentation, something as simple as opening up an image in a tab requires a Developer Extension.   Adding a User Extension system to JupyterLab -- ideally, one that is close enough to the Notebook Extension system that popular Notebook Extensions can be ported to JupyterLab relatively painlessly -- would go a long way towards addressing the concerns that people have raised here.

There are several possible paths to achieving this in JupyterLab:

- Allowing JavaScript inside Notebooks, as was done in the classic interface.  This is a big decision, warrants discussion on its own, and the main JupyterLab developers have certainly thought about this a great deal.  
- Developing generic extensions that can be rapidly customized through settings and/or small (and interpreted at runtime) JavaScript scripts.  I believe that there is some work in the JupyterLab community going on already on this.  We would be happy to contribute to this.
- Make JupyterLab at least partially programmable through a scripting interface -- the obvious one is JavaScript.  Again, the JupyterLab development community should weigh in on this, because there are certain to be concerns  which need to be discussed and addressed in a thoughtful way.

In my 2+ years of using JupyterLab, I've found it to be (overall) a big improvement over classic Notebook for my purposes.  The ability to run other executables in tabs of JupyterLab has many, many use cases.  I have shared the experience of many here that it's much harder to customize than classic Notebook, and if one outcome of this thread is an easier-to-customize-and-extend-by-end-users JupyterLab that would be very positive for everyone.

**Comment by fperez** [comment]
Hi all, (long post, sorry! tl,dr; skip this comment and read the next).

First I want to speak a little on behalf of the project to thank not only Zach for opening this issue, but Zach, Darian, Kevin and others who have been putting a lot of effort into the maintenance of the Classic nb repo over the last few years We have no dedicated funding right now for this work, and their efforts are immensely appreciated.

Next, I want to also communicate that the voice of users who have brought up their concerns here is _critical_, many thanks for your input. No, the project will _not_ abandon you or your work, and leave you facing the impossible task of rewriting years worth of tutorials, educational materials, etc.  We need to find a path forward that can bring everyone along, but that is also sustainable for our developer team, which is incredibly small considering the scope of impact of the project.

There seems to be good consensus above on the need to transition away from the Classic _codebase_, but clearly the community needs the Classic _user experience_ to remain available for certain use cases. Importantly, that experience must be preserved such that:

* Key extensions that users depend on for their work remain available. nbgrader, RISE and TOC have risen to the top of this list, but we need to find out whether there are more.
* Existing educational materials, screenshots, videos from previous classes, etc, remain valid and can continue being used.  New solutions don't need to match Classic pixel-wise, but they _can not invalidate the content_ people have developed and depend on for their work. UI layout, menu structure, keyboard behavior, etc. need to match Classic in a way that allows folks to experience no disruption to their habits or use of training materials.

Now the reality is that a skeleton crew like ours can't maintain forever a codebase that has been described as a spaghetti nightmare of global state, and we also have evidence (in the thread and from my own experience with a very large class of 1,200 students) that Lab can also work well in educational scenarios. This does _not_ invalidate the concerns raised above, it only shows that Lab can be a good option in some cases that include education.

So the question is, how do we meet these multiple needs?

At last Friday's governance meeting (which is always open, though mostly boring and these days typically devoted to drafting our final PRs :), a number of folks came to discuss this, and Darian presented a [short slide deck](https://docs.google.com/presentation/u/0/d/1w7Xp_7pbVz8EqQ9gyytB7CT2EEAJox8sqYieGWRYHI4/edit) with a plan, that summarizes what @gutow and others above have already outlined here. In brief, a way forward would be:

* Continue our planned gradual sunset of the Classic _codebase_, but providing what maintenance is possible (and we welcome assistance on this front!) to ensure that anyone who can not change anything other than a minor version number of notebook 6.4.x can continue working as is. But notebook 6 would be the last version _based on the Classic codebase_.
* Build on the excellent work by @jtpio building [RetroLab](https://github.com/jupyterlab/retrolab) to offer the Notebook _user experience_ as Notebook 7. Specifically this means that the notebook v7 release _would be retrolab_ (yes, this would involve some repo wrangling, details later).
* Notebook 7 would be released only after ensuring that it really meets the needs of current users as stated above. Critically, this means putting in the necessary work to port/adapt critical extensions like nbgrader, RISE and TOC to work on this future nb7 from Day 1.
* Notebook 6 would remain in minimal maintenance mode at least for a while. Details TBD.
* Do outreach to identify any other critical extensions that might potentially block large communities from upgrading to nb7, and similarly add them, if possible to the above list of "day 1" ported extensions.
* For the "long tail" of other extensions, provide a number of resources such as tutorials, assistance on gitter/github, online sprints, etc. to ease the porting effort.

During the meeting we hashed the above into a slightly more detailed set of notes. I'll post them in the next comment as a "proto JEP" so we can collect early feedback here, and once there's a bit more input we can then turn that into a properly written JEP to organize and coordinate effort.

I am an educator too, though I don’t teach Data 8 at Berkeley, which has been discussed here. But I do want to flag that for Data 8, @yuvipanda has already opened [an issue](D8 - https://github.com/berkeley-dsep-infra/datahub/issues/2422) where we are tracking any blockers to try out using RetroLab as the default interface for future iterations. I personally teach Data 100, the follow-up course to Data 8, and right now I've been using JupyterLab as the default UI. I worked with [Yuvi and the team](https://github.com/berkeley-dsep-infra/datahub/issues/2606) to make the transition early on, and after a few minor bumps it's been pretty smooth sailing, as far as I can tell (at least no problems have bubbled up to me directly).  I only say this to indicate that even on courses like the big Berkeley ones, we're looking at ways to move forward on top of the Lab architecture, but that may include using Retro for some of our introductory offers where the Classic UX is more appropriate.  But I also mention it because I've seen during office hours students (most of whom are not CS majors nor professional programmers) taking active advantage of Lab's features like multiple notebooks open, text editors, CSV and other data browsers, etc.  This is not to dismiss anyone's concerns, but just to provide illustration that Lab _can_ work very well in certain educational settings, even for introductory, non-programmer audiences.

Now, I finally want to address one more point that has been brought up: the ease of playful exploration with extensions that was offered by the Classic system, which is not there in the more complex Lab architecture. This is true (I certainly find the Lab web stack daunting), but I hope that we can also develop not only supporting material to ease the experience, but also explore whether it's possible to expose for example a (perhaps limited) simplified extension API that meets important use cases while being very easy to use. We have a great precedent of this pattern in something like seaborn, which for many common statistical plotting needs can shield users from the complexities of matplotlib, while leaving it available for those who need it. I would love those with more experience with TS/JS and interest in this problem to pitch in with ideas on this front.

This post is already very long, so thanks for reading. I hope I've been able to communicate at least that a key concern for many of us on the team is to _not make our users feel abandoned_! To the extent that may have happened and caused distress to folks above, like @psychemedia, @gutow, @jph00, @labarba, and others, I'm sorry! We absolutely treasure everything you've done in this community, the project exists to serve users like you, and we want to find solutions that serve you well.

I do believe that the architecture of JupyterLab is a fantastic foundation on which to build the kinds of experiences we want for interactive computing and data science. But the key term is experience**S**: there can be more than one, and they can serve different valid needs. 

 A few closing points:
* I want to thank @jhamrick and @damianavila for excellent feedback on what it will take to port nbgrader and RISE. This will not be trivial work, and we would very much love to find help and resources to speed it up.
* I want to thank @jtpio for his fantastic retrolab. We have a solution to build a path forward thanks to this work (I've begun [filing](https://github.com/jupyterlab/retrolab/issues/259) [issues](https://github.com/jupyterlab/retrolab/issues/260) to identify important parity mismatches, please go ahead and pitch in too!).
* I am actively working to find resources to better support the scientific and educational community in Jupyter (my own space). I hope if some of those efforts are successful, in the future we'll be able to handle situations like this with more foresight to avoid some of the headaches that have occured.

Thanks for reading, and for all you do for Jupyter!

**Comment by fperez** [comment]
# ProtoJEP - the evolution of the Classic Notebook v6 and v7

-   Notebook 7 is based on the JupyterLab codebase, with "the same" user
    experience.

-   When Jupyter Lab 4 or notebook 7 are installed, it's a single set of
    tools, and both commands start the majority of the same machinery.
    Typing 'jupyter lab' or 'jupyter notebook' will allow users to move
    between the two visual experiences easily. Installing one always
    installs the other so the underlying tools are always available.

-   There will be seamless ways for users to move back and forth between
    notebook and lab in those applications.

-   The \`jupyter lab\` and \`jupyter notebook\` commands always start
    both and only determine the default view.

-   Define "the same"

    -   Existing documentation works unchanged: old screenshots, videos,
        manuals, keyboard shortcuts, etc, that explained how to do
        something in classic, are not invalidated. Educators who have
        such materials online don't need to change anything.

    -   Critical extensions should work on day 1. We need to identify
        those; These extensions will need updates to work with the new
        API, but we pledge to work on porting those core extensions.

        -   Nbgrader

        -   RISE

        -   What else? We need to find out.

            -   Matthias: TOC ? Port it to reuse the current Lab TOC,
                but with its own, notebook-native UX?

    -   Work has been done to document and assist authors in porting
        their own extensions, so they might have updated them to nb7.

-   There will be a central location with information on how to port
    extensions for others to do so in the future.

-   Performance?

    -   What do we want to tell users about or commitment to
        performance.

    -   Measure and communicate: most likely, performance will be equal
        or better than in classic, as the bulk of the slowdown in Lab
        comes from multiple tools on the page and activity switching. By
        definition those won't exist in nb7. But we need to have some
        hard data on this to communicate.

-   Future goals:

    -   Improvements to the APIs to facilitate exploration in JS/CSS.
        Some of the ease of lightweight exploration of the web page that
        is possible under nb6 also creates the maintainability/security
        nightmares we have. How to create most of that experience under
        the new APIs will need development.

    -   Potential new high-level APIs for a limited class of extensions,
        akin to how Seaborn provides limited, but powerful and
        easy-to-use plotting in matplotlib for a set of common use
        cases.

User Stories (north star, long term)

-   As a user, when I install lab or notebook, I get both experiences,
    so I can choose between them as I work.

-   As a user, when I start the application using \`jupyter notebook\`
    or \`jupyter lab\` the default view that I see will correspond to
    notebook or lab respectively. The ‘’handler is also updated
    accordingly to point to that application.

-   As a user, when I am using the lab or notebook interface, I can
    easily move back and forth within the UI using easily discoverable
    menu items, context menus, buttons, etc.

    -   Open this document in the notebook interface

    -   Open this document in the lab interface

    -   Open this directory in the notebook file browser

    -   Etc.

-   As a user, when I am using notebook 7, the visual design strongly
    matches what I am used to for the classic notebook.

    -   As a user, I want the visual design of the notebook interface to
        get better over time, so my user experience improves over time.

    -   As a user, I want the visual design of the notebook and lab to
        be closely aligned, so I know that I am using a single, unified
        experience.

    -   As a user I want to UI to not change dramatically so that my
        tutorials, and class material do not get out of date.

-   As a user, I want extensions that need a left/right/bottom/status
    area to work in notebook, but I want those areas to be entirely
    hidden by default, so I can have the document oriented notebook
    experience that I am used to.

-   As a user, I want to know all of the great new things I am getting
    with notebook 7, so I can make an informed decision to upgrade.

-   As a user, I want the most common nbextensions to work in both the
    notebook and lab interfaces, so I can move back and forth without
    missing functionality.

    -   RISE

    -   Nbgrader

-   As a user, when I install *Jupyter Desktop*, (the lab4/nb7-based
    native desktop suite, taht consists of the *Notebook* and *Lab
    *apps), I can double-click on an .ipynb file on my local computer
    (Mac, Win, Linux) and that file opens as a single document that I
    can edit, run code into, etc. The same installation of Jupyter
    Desktop also offers a JupyterLab application with a more IDE-centric
    experience, including file management, terminals and more.

    -   From my command-line I can open single files with either
        application (Notebook or Lab), and I can also open Lab pointed
        to a whole directory.

    -   In the future (not planned as part of the initial release),

        -   Jupyter Desktop will also let me open my Jupyter editor
            (with any extensions I may have installed) as a single
            application, and Jupyter Console.

        -   This will let me open otherJupyter-based tools in my native
            desktop experience, such as Consoles attached to Notebooks,
            in their own Native window.

        -   Jupyter Desktop will recognize and syncrhonize with any
            web-based Jupyter Servers I may have opened on the same
            system, and can also connect to remote servers and kernels,
            so that I can have the same native Desktop experience
            against remote resources.

Technical details

-   Nb7 is the current retrolab code bases.

-   Notebook repo is moved (at some point) under jupyterlab org.

    -   Keep on the same org to allow to migrate relevant issues

-   Port nbgrader and RISE to work with lab4/nb7.

-   Question: should the notebook repo get a PR for nb7 with retrolab
    code, or move the retrolab into notebook

    -   Likely want to rename “retrolab”repo, to “notebook”and migrate
        relevant issues to it. We can rename "notebook" -> "notebook6"
        and leave it permanently available but with no new work done on
        it after a final release.

    -   Will need to sort out details of issue management - technically
        it would fit better under jupyterlab, but the jupyter/notebook
        name/location is great, and moving

Community and communication details

-   Find out which other extensions have a critical impact for large
    user bases.

-   Gather feedback on github, discourse, etc before turning this into a
    JEP.

-   Engage with the Jupyter Extensions community.

-   User-facing communication on nb7: it's a brand new version of the
    notebook that has the following features:

    -   Theme less is more:

        -   Familiar

        -   Nostalgic

        -   Reviving a beloved project/technology

    -   Real-Time Collaboration

    -   Dark mode

    -   Debugging

    -   Internationalization

    -   Accessibility. (we won’t have accessible experience, but theres
        less work to do)

    -   Access to the Jupyter Extensions Store (made-up name)?

-   We should host clinics for some amount of time to help plugin
    authors (nbextensions) to modernize and port their extensions over
    to nb7.

**Comment by meeseeksmachine** [comment]
This issue has been mentioned on **Jupyter Community Forum**. There might be relevant details there:

https://discourse.jupyter.org/t/governance-office-hours-meeting-minutes/1480/165

**Comment by jph00** [comment]
> Existing educational materials, screenshots, videos from previous classes, etc, remain valid and can continue being used.  New solutions don't need to match Classic pixel-wise, but they _can not invalidate the content_ people have developed and depend on for their work. UI layout, menu structure, keyboard behavior, etc. need to match Classic in a way that allows folks to experience no disruption to their habits or use of training materials.

This is fantastic @fperez, thank you so much. That would be a really wonderful outcome.

The one extension I'd really love to see included (and is loved by much of the fast.ai community too) is collapsible headings, including the ability to use arrow keys to jump to the top or bottom of a section (or open/close a section when a heading is selected). Without this, keyboard navigation of hierarchical notebooks is challenging.

**Comment by Zsailer** [comment]
Hey folks, so grateful for the great discussion here. Thank you to all who have participated! :heart:

I've created a new issue (#6220) referencing @fperez's ProtoJEP comment above. Please use that new issue to give feedback specifically on that (pre)proposal.

**Comment by fperez** [comment]
@jph00 have you seen how JLab currently handles this?

![jlab-collapsible-headings](https://user-images.githubusercontent.com/57394/140166118-bac270da-ba60-48e6-bef1-9dc80b997aec.gif)

Do you think this meets your needs, or is there more that should be done/changed here?  

Thanks again for your awesome feedback (and work on Jupyter-related tools over the years!!)

**Comment by fperez** [comment]
@jph00 for one thing, I _think_ the JLab version is only mouse-based. I'm not familiar with the details on how this works keyboard-wise in Classic, but perhaps we can open an issue to track this specific point. I'm a huge fan of good hierarchical document management -- the LyX navigator/outliner is IMHO still better than many others, as it also allows for safe structure-based reordering of the document. So improvements to this that help all users would be fantastic (on Retro/Classic but _also_ on Lab!).

**Comment by jph00** [comment]
Yes that's right, lab doesn't have the keyboard shortcuts, which is the key thing - since with the keyboard shortcuts one can jump to start and end of sections so easily, collapsing them actually becomes much less important. Note also that collapsible headings integrates with the TOC2 extension to allow them to optionally sync collapsed sections.

Here's what the docs say about the keyboard shortcuts:

* Command-mode keyboard shortcuts, (enabled by default, and set to left and right arrow keys to collapse/expand sections, or go to the previous/next heading, plus shift-right to select a heading cell's section, shift-a/b to insert a heading above/below the current cell, ctrl-shift-left and ctrl-shift-right to collapse/uncollapse all headings). Bindings are also configurable from the config page

Details: https://jupyter-contrib-nbextensions.readthedocs.io/en/latest/nbextensions/collapsible_headings/readme.html

**Comment by fperez** [comment]
Thanks for that clarification @jph00 - I added that to #6220 so we track it more carefully and it doesn't get lost in this very long thread.

**Comment by afshin** [comment]
I opened [an issue in the JupyterLab repo about the TOC extension's shortcuts](https://github.com/jupyterlab/jupyterlab/issues/11401) to track this.
(cc: @jph00 @fperez)

**Comment by jgwerner** [comment]
Re: @Zsailer 
> Another thing I've heard in this discussion is that there are key/critical notebook extensions on which many users depend, e.g. [RISE](https://github.com/damianavila/RISE), [nbgrader](https://github.com/jupyter/nbgrader), [pixiedust](https://github.com/pixiedust/pixiedust), etc.
> 
> What is the story for users who depend heavily on such extensions?

We use `nbgrader` quite extensively. I believe it would help to clarify [that two epic UI packages included with](https://github.com/jupyter/nbgrader/tree/master/nbgrader/nbextensions) `nbgrader`:

1. Extensions included within the Jupyter Notebook Classic interface (either within the Notebook itself or in addition to locations in `/tree`). These include the `assignment_list`, `course_list`, `create_assignment`, and `validate_assignment` extensions. These server and client extensions either manipulate the `ipynb` metadata or help fetch/copy files using a directory structure.
2. The [Formgrader](https://github.com/jupyter/nbgrader/tree/master/nbgrader/nbextensions/formgrader) extension provide management capabilities to distribute and collect assignments, provide inline comments, release feedback, trigger auto-grading tasks, etc.

> Note: links provided above are for `nbextensions` only. These rely on the `serverextensions` that use the same names.

The first set of extensions should be easily portable to JupyterLab or any other Notebook interface for that matter and in my personal opinion doesn't require a specific Jupyter Notebook version or setup, in so long as it's possible to develop extensions that can communicate with the file system and that can edit ipynb metadata (file-wise and cell-wise). We are currently porting the client extensions to use the JupyterLab setup but most of our users ("our" is IllumiDesk, we are an EdTech company that leverages nbgrader and Jupyter extensively) prefer the Classic interface so are developing the extensions to maintain compatibility with the RetroLab setup.

We decided on embarking on a complete rewrite of item 2 above (`Formgrader`) using a more modern React-based UI that communicates with [nbgrader's RESTful APIs](https://github.com/jupyter/nbgrader/blob/master/nbgrader/server_extensions/formgrader/apihandlers.py). This was certainly an item that created issues since the `Formgrader` interface is basically a webapp that doesn't need to depend on a Jupyter Server. [The code for React frontend is [available here](https://github.com/IllumiDesk/lms). (We recently moved the package to our internal monorepo however this solution was fully tested with nbgrader v0.6.2). This is a little bit of a technisism but we also developed a custom handler for this new Formgrader (aka [Formgradernext](https://github.com/IllumiDesk/formgradernext)) to server the bundle via CDN, but that's just an implementation flavor 😄 

cc/ @labarba @BertR

**Comment by astrojuanlu** [comment]
Can jupytext be added to the list of extensions that aren't yet fully compatible with JupyterLab? I think having text-based formats (rather than `.ipynb`) was a great step forward to solve the problems of notebooks under version control, but it can't be used easily anymore https://github.com/mwouts/jupytext/issues/271

**Comment by fperez** [comment]
> Can jupytext be added to the list of extensions that aren't yet fully compatible with JupyterLab? I think having text-based formats (rather than `.ipynb`) was a great step forward to solve the problems of notebooks under version control, but it can't be used easily anymore [mwouts/jupytext#271](https://github.com/mwouts/jupytext/issues/271)

Duly noted @astrojuanlu, thx! See the PR above where the actual JEP is starting to take shape.

**Comment by labarba** [comment]
Adding to the [comment](https://github.com/jupyter/notebook/issues/6210#issuecomment-966422757) I was tagged on above by @jgwerner — I use the `nbgrader` extension to create assignments, i.e., to write problems in a Jupyter notebook with cells being tagged as read-only, graded, or test—that is, the [Instructor toolbar extension](https://nbgrader.readthedocs.io/en/stable/user_guide/highlights.html#instructor-toolbar-extension-for-jupyter-notebooks). We don't use `nbgrader` for any of the assignment management tasks, i.e., "formgrader" extension.

I have dozens if not a hundred student assignments written over several years (plus re-write or tweak exams every semester). These are all auto-graded using `nbgrader`, programmatically, in a Docker container in the course platform. Our workflow auto-grades immediately a student's assignment when they upload it to the course, and they get instant feedback of right/wrong answers.

**Comment by choldgraf** [comment]
Hey all - just a note that we're nearing final language around the Notebook v7 JEP, please provide comments and thoughts soon before we head to a vote. As there has been a lot of discussion and refinement already, it'd be most-helpful if comments are focused around specific and actionable things to improve the proposal! https://github.com/jupyter/enhancement-proposals/pull/79

**Comment by psychemedia** [comment]
@choldgraf Just an observation that if you wanted to broaden comment outside the dev community, expecting folk to know how / where to comment on elements of the proposal via GitHub PRs makes huge assumptions about the skill base of the folk whose comments you want to solicit...

**Comment by yuvipanda** [comment]
Wow, that is a lot of work in a short time, @choldgraf! I'm thankful to all the authors who put in work on the concrete proposal!

**Comment by jgwerner** [comment]
@choldgraf I'm not sure if this the right channel or not for this offer, but we would be more than willing to sponsor the development of the `nbgrader` extensions for Notebook v7. I also assume (please correct me if I'm wrong) that before writing any code that there is a process to validate the UX? Or do you envision more of a "lift and shift" approach based on the current version of the `nbgrader` `nbextensions`?

We have the `Formgrader` web app (currently built as nbextensions) ported already to a React-based solution but perhaps we tweak that package a bit to make some of the assets configurable (logos, theme, etc).

cc/ @labarba

**Comment by choldgraf** [comment]
Hey all - I'm going to close this issue, as the JEP in https://github.com/jupyter/enhancement-proposals/pull/79 was approved, which means that our immediate term plan will be to base Notebook v7 on JupyterLab components, with the goal of making the package more maintainable and also bringing in new features that have been created in the JupyterLab ecosystem.
